### PR TITLE
New customizable editor theme

### PR DESCRIPTION
**Pull request by volzhs**
Demonstration

godot3_editor_theme_final

Note that, 2.x old theme is used for creating UI for game.
This PR only affects editor theme.
And applied some of image resources from @djrm (thank you)

edit : I think there are more icons should be adopted from what @djrm did.

PS. could not test with visual script part, I couldn't create visual script at the moment.

### CODE REVIEW

**Comment by toger5** [comment]
Can imported custom editor themes also use those variables or is it only for the default theme?

**Comment by volzhs** [comment]
@toger5 it's only for default, because it's specially made with `StyleBoxFlat` to apply color with contrast.

**Comment by djrm** [comment]
@volzhs why did you changed the theme that much, i like the customization concept, but you did a lot of changes to the new theme (without community discussion) you should have leaved the theme as it was and then we could improved if need it, the new theme that you proposed, looks in my opinion really strange and incomplete, it still uses old elements that dont fit, and look like a windows 10 program with really sharp edges.

or maybe i dont understand what are you doing, is that another theme that is not supposed to be the new default?

**Review by djrm — CHANGES_REQUESTED** [review]
new widgets must be discussed

**Comment by volzhs** [comment]
@djrm Yes, some art resources are not adopted yet.
The reason why I did it like this is that we can't changes colors for each UI elements with textures.
(you know, some people said about color, now they can get whatever color they want with this)
I hope you to understand for it.
I would like to contact to you before this PR, but for personal reason, I was in hurry. (can't do much for next 5 days)
I'm sorry for that.

**Comment by djrm** [comment]
@volzhs i understand that people want to customize the engine color, but first thing it can be done with textures (at least certain parts of it), and second i think that the people that are going tho change the color of the engine is minimal (it was not worth the effort, because if other "alternative" colors were needed the it could be as simple as creating a new theme variation, for example with grey background) it is simply not worth it to restrict the themes to be simple boxes in favor to color customization that just a few people will use.

**Comment by toger5** [comment]
How hard would it be to add modulate colors for textures and provide a couple of default variables: background color, foreground color, highlight color to use in a theme (.tres and default_theme)
So a style box has an addition property: `modulate color` which than can be set to `<background_color>`
or a flat style box can use `<highlight_color>` as the border color...

This would be some additional work but would allow for tremendous customisation!!
We could use djrm's stylboxes which are really nice and we just change the color to white/ different grey values and give them a modulate color based on the exposed variables.

That would be an amazing addition to the editor in general and would also apply for any user created theme. 
@volzhs what are your thoughts on that? How hard would you estimate is it's implementation?

**Comment by djrm** [comment]
@toger5 i was doing that when i was creating the theme, the only problem is that sometimes you need to modulate using to colos, for example in inputs, the background needs to modulate to the bg color and the line needs to modulate to the highlight color, i thing it could be solved by rendering two textures or using a custom shader. but im not sure if it is worth the effort, as i said we could offer 3 variations, i really dont think it is good to offer the full range of colors i mean we will do this just for someone who wants the engine to be green and purple. most people eanted gray-ish variation, we could offer dark-grey, standard, and light-grey

**Comment by Zylann** [comment]
I think the main point of this PR that changing theme colors, or making a theme in general, is a tedious task (getting a copy of the base theme as source, modify each SVGs, export them, install theme, hope it works).

@toger5 modulate is not always enough, you might need more core colors, and it requires the base textures to be black and white otherwise colors will badly blend.
As I said in another issue, making/modding a theme in a modular way (with .tres) should be made easier rather than trying to add special customizers for the built-in editor theme.

**Comment by volzhs** [comment]
@djrm 
why I reverted your commit for new theme was that it changed default theme a lot as I said before.
old default theme is very useful when make prototype.
this is most important reason.
so I needed to revert it, could not change one by one.

I know and feel what you have done is great.
but it does not provide proper visual than old one for general.

how about finding a way to look better, based on this?

**Comment by volzhs** [comment]
@djrm I will change icons and some UIs you did on last commit when I have time.

**Comment by akien-mga** [comment]
To rephrase what @volzhs as been saying, and from what I understood when discussing it on IRC with him, the motivations for this PR are:
- to restore the default_theme which is used in games, as it's more functional than the new version we had to produce the wanted editor theme
- to decouple editor themeing from default_theme, so that we can fully customize the editor theme without having to modify default_theme

To do so, @volzhs started by reverting @djrm's commit which modified default_theme, and has then worked on reimplementing elements of @djrm's work with a cleaner approach.

So if we agree on this new approach, the idea would then be to work together to implement the new editor theme that we want incrementally. The possibility to tune some colors in the editor settings is not the main point of this PR, but just a nice-to-have tweaking possibility - the main point of this PR is to restart the work on a better foundation.

I hope I summarized it properly :)

**Comment by djrm** [comment]
@akien-mga yeah, i understand that however, the way @volzhs is implementing the color customization is what causes the problems here, since the way that is done is this PR does not allow the  use of textured widgets. i completely agree with having the old theme for game prototyping but sacrifizing the editor theme to have color customization is simply not worth it. maybe

**Comment by djrm** [comment]
also could it be possible to create a TRES theme, that is loaded and shipped in with the engine. that way we can leave the default theme as is, and just pack and load the engine theme.

**Comment by toger5** [comment]
@djrm, @volzhs Although I really love the color customisation I also agree that it is not worth it. There are a couple of grey values which make sense: Very dark (for contrast although it doesn't look that good in the eye of some ppl), blueish grey (because some ppl really like the look and modern feel of that), saddle lighter grey (easier on the eye and commonly used (maybe overall best solution))

to give full color variation will just waste time playing around and only has benefit for the crazy ppl under us who like all green or pink.

After seeing the approach with adapting the background color to the window color (see here: #8588) I still see benefit to allow full color variation for some ui elements. also looking at the discussion for the highlight color I also see some advantage in the full variation. This is why I would love to see the option to give optional access to some colors of a .tres theme and the default theme. But I agree it is not worth for sacrificing the really good looking theme we have right now.

In @djrm current default theme it would be enough to just give access to the three different grey values (*bg*, *dock top part*, *dock panel*). This would solve most controversial and would look amazing.

Everyone who want even more customisation can install another .tres theme.

Just my thoughts

**Comment by Zireael07** [comment]
I am in favor of being able to customize the color as you wish. Saying "just install a .tres theme" is not an option (right now) because .tres themes are known to be difficult to make work. (My opinion will change if they are made more intuitive/less buggy)

Yes, the range of really useful colors is limited. But the customization is what FOSS is about, IMHO.

> not worth for sacrificing the really good looking theme we have right now.

Why do you think we're sacrificing anything?

**Comment by n-pigeon** [comment]
We also have to make better default theme for games prototyping. Something that doesn't have custom editions that were added to make Editor GUI work hehe :)

**Comment by reduz** [comment]
For game prototyping in general you dont use themes.. people prefers the
simpler anchor system and stuff like texturebutton..

**Comment by n-pigeon** [comment]
I think this is more problem of engines that don't have GUI toolkits :D
It depend on type of game, how much custom look is needed, how early prototype is and so on.
GUI toolkit is also useful for in game dev tools and game settings.
Also people use Godot for applications so it won't be wasted :)

Aaaaannnnd people who prefer Texture Buttons all over the place are usually the ones that are not making them. I have friend who quit his job because he was making GUI textures for GUI heavy game for 6 months.
Horror

Good GUI tookit is one of those things I love in Godot.

**Comment by Zylann** [comment]
@reduz although my own game doesn't as well (only sliders and checkboxes) because it's a small game, the GUI nodes should have a basic theme at least, that represents most of common GUI states. (I also feel your pain, @n-pigeon ^^) It can be OK even for prototyping. Also, about "custom edits for editor", it's better done through data, not new classes.

**Comment by volzhs** [comment]
image

Nice round play button group panel is back.

**Comment by volzhs** [comment]
image

more works on slider, tooltip, dock selection icon

**Comment by volzhs** [comment]
image
more works on spinbox, option button, property editor, animation editor

**Comment by volzhs** [comment]
I think it's pretty much done to merge.

**Comment by toger5** [comment]
@volzhs do you know why moving the line between docks (resizing them) is so laggy?

**Comment by volzhs** [comment]
@toger5 it's debug build.

**Comment by n-pigeon** [comment]
You are a magician :D

**Comment by TeddyDD** [comment]
I could set up Godot to match my GTK theme, this is so awesome ;-;
zrzut ekranu z 2017-05-19 12-03-56

**Comment by volzhs** [comment]
@TeddyDD that is better than I expected...

**Comment by akien-mga** [comment]
@volzhs Note that the `editors/` category is for editor plugins. I think the `theme` subcategory would be better fitted in `interface/theme`.

**Comment by akien-mga** [comment]
I tested the two themes (@djrm's, the current merged theme, and @volzhs', from this PR) next to each other to get a better overview. I used the customization feature in @volzhs' version to pick a colour similar to @djrm's, to ease the comparison. Here are some screenshots.

The order is always:
- @djrm's theme (recognizable by the capsule-shaped input fields and rounded buttons)
- @volzhs' theme (recognizable by the lack of those, and thus blocky input fields and buttons)

## Project manager

09_djrm
09_volzhs

## 2D editor

01_djrm
01_volzhs

## Scene menu

02_djrm
02_volzhs

## Assetlib

03_djrm
03_volzhs

## Create new node

04_djrm
04_volzhs

## AnimationPlayer

05_djrm
05_volzhs

## Editor settings

06_djrm
06_volzhs

## Script editor + profiler

07_djrm
07_volzhs

## 3D editor + audio

08_djrm
08_volzhs

**Comment by n-pigeon** [comment]
Yeah. I would still tweak some stuff to make it more like djrm's theme. :)

I see many people really don't like this green. I think this green is great, because it makes stuff pop and looks modern. 
Blue highlight blends too much with the backgrounds in my opinion :P

**Comment by djrm** [comment]
@akien-mga  i guess we can't have both versions, i mean if this PR gets merged, the theme will have to change completely, i made my theme to be consistent with the overall icons baseline size and other icons, like arrows (currently the theme in master is a modified version by volzhs and it doesnt look bad but i explained my opinion against no borders in the vector theme issue, but as i suspected they look bad) for example this, it will be extra effort to add margins where needed.

screenshot from 2017-05-19 13-22-45
screenshot from 2017-05-19 13-23-31
screenshot from 2017-05-19 13-24-05

i dont want to offend anyone, but the implementation in this PR looks really cheap and inconsistent, i think it can look better but a special set of widget icons need to be created for it, and fix some widgets, my style and the stylebox flat can not be used together, so we have to decide, if we keep the theme in master and fix its remaining issues or we merge this and do more work to make it look better (or at least consistent), the other option is to merge this and ignore how the engine looks, i think not much people actually cares about consistency or how things looks (ignoring the fact that every person has a different taste and quality standards), in any case people can use custom themes if they want so i guess any choice is fine, i will be looking to what happens and try to help in what i can.

so my last opinion about this is:
- if people want to have customizable colors lets have them (even though i think it is not worth it)
- i agree with having the 2.x theme as game prototyping theme.
- my theme and the proposal in this PR are not compatible so if this is merged, we will have the options mentioned above.
- if my theme is used we will have to fix the remaining issues.

**Comment by akien-mga** [comment]
So, I've given this topic some thought... It's not easy to have two competing themes and to have to select one; even though I see both proposals as WIP and needing more work. The main question for me is what is the sanest base to build upon together.

First of all, a BIG thanks to @djrm for the very creative theme you designed, I love the overall look and feel. And a BIG thanks to @volzhs for trying to address some of its shortcomings while reimplementing many of its ideas. You both did a great work and I hope you'll continue to work together to improve the theme little by little up to the final 3.0 version.

After a few weeks of testing of @djrm's current WIP, and after having tested @volzhs' PR, I would advise that we go with @volzhs' proposal for now (i.e. merge this PR), and then work on reimplementing some of the good ideas from @djrm little by little (I love the capsule shaped input fields and the rounded buttons).

The main advantage of @volzhs' implementation IMO is that:
- it restores the default_theme (you can see in the above screenshots that the default font for prototyping was way too faded for example). I think we can still rework the default theme for 3.0 if we have enough time and a good proposal, but for the time being it's good that we stick to something that works. Theming the editor is enough work without having to care about all Godot games using default_theme.
- it improves contrast (and even makes it configurable). The currently merged theme looks great, but many elements lack contrast IMO and lower the overall usability.
- the changes are less intrusive, so (for now) there is less risk of weird color combinations in some parts of the editor UI

On the other hand @volzhs' current proposal lacks polish, some elements would have to be offset by a few pixels, etc., to look better, and I think that there are various textures from @djrm's theme that should be readded little by little. As I mentioned, I really like the rounded elements for the input fields, buttons, etc.

If the color and contrast customization feature works, I think it's a nice addition. It's not a strong requirement, but as long as it offers some degree of freedom, users will like it. It shouldn't prevent us from adding textured widgets; if we can't modulate them to *always* look great with any customization, it's fine. The customizer will have to stay reasonable in their color choices :)

To summarize my opinion:
- Design wise, I really like what @djrm aimed for, but it has several drawbacks which are hard to address due to it heavily modifying the default_theme
- @volzhs proposal is less intrusive, and though unfinished and unpolished, I think it's a good base for us to work incrementally on improving it. The final theme could (will) look different from what @volzhs currently implemented, and I'd be totally fine with going closer to @djrm's idea
- So I would propose to merge this PR, and then work on polishing it and refining widgets one by one in dedicated PRs, instead of the big change-it-all PRs we had so far for both proposals. So one PR that would reimplement the nice capsule-shaped input fields, one for buttons, etc.

WDYT?

**Comment by reduz** [comment]
My thoughts are as follows. I think we should go with @volzhs solution because this way we can offer paying customers a Godot version with a darker theme.

No, actually, here's my thoughts about the whole thing.. I mentioned these things before so it's no news.

1) I don't like round corners. I think they make Godot look old, as current trends go all towards avoid them. Still, if you guys are OK with them, feel free to leave them there.
2) I don't like the lack of titlebar or separators in @djrm 's theme. I think it's ok that this can be achieved, but I would prefer Godot uses a less experimental theme approach, so users know the stuff they want to use exist.
3) I like @volzhs's approach of customizing the theme colors. Users requested this consistently, and I guess It's good they can personalize the editor as they like without having to write a theme from scratch.
4) I think @djrm's theme is more polished, so if you guys can work together, that would be best. Otherwise, I'd prefer we go with @volzhs 's work too.

Thanks both for your awesome work!

**Comment by djrm** [comment]
@akien-mga, @reduz ok, lets go with this then, i dont know if  i will be able to come up with a compatible idea with this but im looking forward to that.

**Comment by akien-mga** [comment]
Alright, I've merged this PR as discussed. I've also branched off the previous commit in the *djrm-theme* branch - that branch is not meant to receive new work (the theme changes should be done in master based on the new basis), but it can be useful to retrieve the icons in `editor/icons` that are not yet in @volzhs' WIP.

**Comment by reduz** [comment]
I'm sorry, I think I broke the scrollbar and slider logic, as this conflicted with the InputEvent changes I was working on. I tried to merge properly, but I'm confident I screwed up, so please give scroll_bar.cpp and slider.cpp a check to make sure it does what you intend.
